### DISCUSSION TITLE
Mapnik 問題

### DISCUSSION THREAD

**Discussion by hfu** [unvt/pm#47] [Ideas]
Node.js で Mapnik を導入することが難しくなってきている。技術的な原因は、パッケージの管理なりサポートの幅の問題であるように見える。

具体的には、Raspberry Pi で mapnik を入れることができない。linux arm64 でのバイナリパッケージがないのでビルドを試みるがビルドされないようになっているようである。

現在のところ、UNVT で Mapnik 依存しているのは charites と itoma であるように見えている。

mapnik を導入するノウハウを確保維持するか、あるいは charites と itoma の mapnik 依存性を外すことが良いのではないか。

なお、Raspberry Pi で docker.io を使って docker で node を動かしても、やはり同じようなエラーメッセージで mapnik 導入がうまくいかない。

**Comment by hfu** [comment]
https://github.com/unvt/charites/blob/c047b6638590116b9ee73699a3ce71934e517914/docs/source/install/recommended_environment.rst にも、arm64 で mapnik が動かないことが言われている。また、sprite-zero が mapnik 依存の要因であるというような情報になっている。

**Comment by hfu** [comment]
https://smellman.github.io/vector_tile_book/docs/knowladge/sprite_url.html も参考になる。spritezero を代替できれば、UNVT は Mapnik 依存を外せるということだろうか。

**Comment by smellman** [comment]
Spritezeroの代替としてImageMagickなどを活用できないかと考えているのだけど、イメージが思ったとおりに出力されるかをどう検証するか？という点で悩んで実装していませんでした。すくなくともSVG->画像へ綺麗に出力される必要があり、有用なツールとして仮想的なHTMLを構築してplaywrightなどでスクリーンショットを撮るという手もあります。

**Reply by smellman** [reply]
おっと、playwrightはRaspberry Piで動かないんだった。

**Reply by hfu** [reply]
ありがとうございます。

ちなみに、https://github.com/microsoft/playwright/issues/7723#issuecomment-964845514 を見ながら Raspberry Pi OS + Docker (ubuntu 20.04) + nodesource の Node で今ちょっと試してみたところ、Playwright が少し動きそうな感じがしました。

**Comment by hfu** [comment]
Docker を注意深く使えば Playwright ならば Raspberry Pi OS でも動かせるかもしれないと感じることができたのですが、同じ環境で mapnik を導入しようとしても、うまくいかないようです。Raspberry Pi OS と同じく linux arm64 扱いとなるのですが、linux arm64 には mapnik が対応しそうにありません。

mapnik 依存性を charites や itoma から外す方向で検討を進めると良さそうな気がしています。

**Reply by hfu** [reply]
ちょっと調べてみたところ、依存性は itoma -> charites -> spritezero -> mapnik となっているので、次の通りなのだと思っています。

- UNVT から mapnik 依存性を外したければ、charites を直せば良い。
- charites から mapnik 依存性を外したければ、https://github.com/unvt/charites/blob/12846cad40c13c7c869314dfba4748b48c88cc99/src/lib/build-sprite.ts にある spritezero 依存を外せば良い。そもそもこの機能を外しても良いかもしれないし、spritezero を直すか代替するかすれば良い。

**Reply by hfu** [reply]
spritezero から Mapnik を使っているところは、https://github.com/mapbox/spritezero/blob/412b5881b1abd54b236056dc3e5b105514ceacc1/lib/generate.js だけのようですね。ここをリプレースできればいい感じになりますね。だんだん、@smellman の理解に追いついてきたかもしれません。

**Reply by JinIgarashi** [reply]
@hfu スタイル編集という観点からするとスプライトを作る機能はcharitesには不可欠と思います。ただ今のbuildコマンドの実装は非常に中途半端で使いづらい(プロジェクトルートにiconsフォルダがあると意図せずに空のスプライトが勝手にできたりする)ので、一時的にスプライト関係の機能を外して、再整理しても良いと感じています。

**Reply by hfu** [reply]
@JinIgarashi 不可欠であるとすれば、Mapnik 依存のない spritezero のフォークを作るというのが長期的な解決策になりそうですね。それまでの短期的な解決策として、一時的にスプライト関数の機能を外すというのも魅力的です。スタイルを編集する人のうち、スプライトを自分で作る必要がある人の数はそれほど多くない、というのが私の経験の範囲内からの推測です。ただ、スプライトを作る機能が不可欠であるかどうかと言えば不可欠なので、spritezero-cli であるにせよ charites の一部であるにせよ、Mapnik 依存がないなどしてポータブルなスプライト作成機能を持つこと、が本質的な課題であるのだろうと思います。

**Reply by JinIgarashi** [reply]
spritezeroをフォークするとしたら、こちらがsfdに対応してるのでいいかもしれません。
https://github.com/elastic/spritezero

**Reply by JinIgarashi** [reply]
elastic版はfontnikを使ってるのか…fontnikのビルド問題が出てきてしまうかもしれません。

**Reply by hfu** [reply]
ありがとうございます。とりあえず、Raspberry Pi OS で elastic/fontnik が入るかどうか試してみました。

![ 2022-04-25 11 52 57](https://user-images.githubusercontent.com/18297/165015524-55193163-689e-469e-9b27-8045c844e52c.jpg)

見た感じ、無事に入っているみたいで、その点では有望かもしれないと思っています。

**Reply by hfu** [reply]
とりあえず、spritezero がボトルネックであって、spritezero を mapnik 依存しないように改造するのが長期的正解である、という認識を手に入れたということをここまでの議論の結論とし、私自身はここで一旦停止します。
